Lib/test/test_http_cookiejar.py

```diff
@@ -123,6 +123,11 @@ def test_http2time_garbage(self):
                               "http2time(%s) is not None\n"
                               "http2time(test) %s" % (test, http2time(test)))
 
+    def test_http2time_redos_regression_actually_completes(self):
+        # LOOSE_HTTP_DATE_RE was vulnerable to malicious input which caused catastrophic backtracking (REDoS).
+        # If we regress, this test will take a very long time to succeed.
+        http2time("1-c-1{}!".format(" " * 65506))
```

Comment:
Would it be possible to use time.perf_counter() to detect if the function takes way too long? For example, test with 1 space, than test with 65K spaces, and test that the second test is not like 1000x slower?

In general, I dislike tests which rely on time, since they are likely to fail on our slowest buildbot. But I don't suggest to use hardcoded timing, only to compare timings of two function calls. It may work on buildbot, especially if the threshold is large enough (1000x slower? more?).

cc @pablogsal

Reply:
With the fixed regex the test should take a fraction of second.

Reply:
@vstinner Do you mean something like the above https://github.com/python/cpython/pull/17157#issuecomment-554333580 ?
The issue here is whether we really should have a unit test for algorithmic complexity of a function at all and what it should look like. Do we do that anywhere else?
I agree that having the one-line test which never explicitly fails is a bit weird.